src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs

```diff
@@ -210,6 +217,12 @@ private async Task ProcessReceives()
                 input.Advance(bytesReceived);
 
                 var flushTask = input.FlushAsync();
+                if (!_waitForData)
+                {
+                    // If not waiting for data allocate another buffer immediately after flushing,
+                    // to try and aquire some before the reader takes the lock.
+                    buffer = input.GetMemory(MinAllocBufferSize);
```

Comment:
Given that the flushTask should always be completed in the benchmark we're looking at, why do you expect calling GetMemory() here instead of the top of the loop to be substantially different? Is it so expensive to retrieve the FlushResult and check few more bools that it's likely to change the timing of GetMemory() relative to another thread waking up?

Reply:
I mention in https://github.com/dotnet/runtime/pull/36992#issuecomment-634121470 that its basically running a data race; in combination with PR https://github.com/dotnet/runtime/pull/36992 it just needs to set the `IsWritingActive` flag near the start of `GetMemory()` before the Reader takes the lock to avoid the lock contention.

I'm not sure it will work and if not then PipeWriter would need a `FlushAsync(moreData = false)` overload; where it doesn't switch out of writing when `moreData` is `true`, akin to the linux [`send(2)`](https://linux.die.net/man/2/send) flag `MSG_MORE`, [`RioSend`](https://docs.microsoft.com/en-us/windows/win32/api/mswsock/nc-mswsock-lpfn_riosend)  `RIO_MSG_DEFER` flag; or [`WSASend`](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsasend#using-dwflags) `MSG_PARTIAL` flag; which would allow it to either keep using the existing `buffer` or request a replacement larger one without having the data race.

Reply:
`FlushAsync(bool moreData = false)` would be better as `Memory.Slice` is cheaper than `GetMemory()` 😉